### PR TITLE
feat: implement new apis and refactoring

### PR DESCRIPTION
**Pull request by zsystm**
##  Implement new APIs
- eth_getCode
- eth_getStorageAt
- eth_newFilter
- eth_getFilterLogs
- eth_newBlockFilter
- eth_getFilterChanges
- eth_uninstallFilter
- eth_getLogs

## Refactoring
- Restricted the scope of access for each namespace API
    - instead of accessing the backend directly, now access is through a scoped wrapper.
- Renamed files and repositioned some methods within the backend directory to better reflect their purpose.

## Fix
- `extractLogsFromEvents` always returns an empty array
    - evm log always emitted with `EventTypeEVM`
    - there are no logs attributes for `EventTypeCreate` and `EventTypeCall`

## Improvements for Future Consideration

- In this PR, we followed the approach used in Ethermint, but it would be better to consider the following improvements in the future:
  - Using an indexer to handle `getFilterChanges` instead of comet websocket connection and pubsub
  - Preventing duplicate log filtering by storing filtered results

## Bump-up cometBFT version
- updated cometBFT version to [initia-labs/cometbft v0.38.x](https://github.com/initia-labs/cometbft/tree/v0.38.x) which includes [this commit](https://github.com/initia-labs/cometbft/commit/6fe9711db1fc4b8a9d17ae41af183e56480baf77)
- without the above commit, broadcasting tx always fails, cannot process any transactions.
  - check the https://github.com/cometbft/cometbft/issues/3261 for details

## License issue
- This PR includes some codes that are derived from the Ethermint, which is licensed under the GNU Lesser General Public License (LGPL) v3. Minievm uses the Business Source License (BSL) 1.1, this could potentially lead to a license conflict.

## Proof of works
**setup minitiad for testing**
1. unzip [minitiad-setup.zip](https://github.com/user-attachments/files/16135675/minitiad-setup.zip)
2. `mv minitiad-setup $HOME/.minitia`
- this setup includes rich account for testing
  - bech32 addr: `init1cml96vmptgw99syqrrz8az79xer2pcgptzssde`
  - eth addr: `0xc6fe5d33615a1c52c08018c47e8bc53646a0e101`
  - private key: `88cbead91aee890d27bf06e003ade3d4e952427e88f88d31d61d3ef5e5d54305`
- erc20 address of genesis coin `umin`: 0xb22eAAbf5a3377abDe553f23eA6d82709BA948E0 
3. clone this branch and run `make install`
4. `minitiad start`

### eth_getCode
**Request:** `curl -X POST --data '{"jsonrpc":"2.0","method":"eth_getCode","params":["0xb22eAAbf5a3377abDe553f23eA6d82709BA948E0", "latest"],"id":1}' -H "Content-Type: application/json" http://localhost:8545`
**Response:** `{"jsonrpc":"2.0","id":1,"result":"0x60806...(skipped)"}`
- `(skipped)`: I wrote ‘skipped’ for convenience, but the original return value is the bytecode of the contract. It was too long, so I shortened it.

### eth_getStorageAt
**Request:**
```shell
curl -X POST -H "Content-Type: application/json" --data '{
  "jsonrpc":"2.0",
  "method":"eth_getStorageAt",
  "params": [
    "0xb22eAAbf5a3377abDe553f23eA6d82709BA948E0",
    "0x386c2fd1f87e34a6f8889cafe74cc4a0d53508a74e2ed92c2375fabcfd8563e9",
    "latest"
  ],
  "id":1
}' http://localhost:8545
```
**Response:** `{"jsonrpc":"2.0","id":1,"result":"0x00000000000000000000000000000000000000000000021e19e0c9bab2400000"}`

### eth_newFilter
- transfer umin to `0xc08d8192da03aafc8c516e35cc4b0cfd2c139f96` before calling this
- above tx must included within scope (fromBlock ~ toBlock)

**Request:**
```shell
curl -X POST -H "Content-Type: application/json" --data '{
  "jsonrpc": "2.0",
  "method": "eth_newFilter",
  "params": [
    {
      "fromBlock": "0x2390",
      "toBlock": "0x2399",
      "address": "0xb22eAAbf5a3377abDe553f23eA6d82709BA948E0",
      "topics": [
        "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef"
      ]
    }
  ],
  "id": 1
}' http://localhost:8545
```
**Response**: `{"jsonrpc":"2.0","id":1,"result":"0xd6f8af542b4b3c699a354336358e4700"}`

### eth_getFilterLogs
**Request:** `curl -X POST  -H "Content-Type: application/json" --data '{"method":"eth_getFilterLogs","params":["0xd6f8af542b4b3c699a354336358e4700"],"id":1,"jsonrpc":"2.0"}' http://localhost:8545`
**Response:**
```json
{
  "jsonrpc": "2.0",
  "id": 1,
  "result": [
    {
      "address": "0xb22eAAbf5a3377abDe553f23eA6d82709BA948E0",
      "topics": [
        "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
        "0x000000000000000000000000c6fe5d33615a1c52c08018c47e8bc53646a0e101",
        "0x00000000000000000000000025051c50bb6f302a7bd89100bead641e65da743b"
      ],
      "data": "0x0000000000000000000000000000000000000000000000000000000000000457",
      "blockNumber": "0x2398",
      "transactionHash": "0x80dcea3db9c181748bce112c760102b2f4d6ac7c25d514ec4ed16656b8fefdb6",
      "transactionIndex": "0x1",
      "blockHash": "0x6ebdf8f6a2ccf6c9b9aead31589524949e8e63c16e938f07100725e02c219432",
      "logIndex": "0x0",
      "removed": false
    }
  ]
}
```

### eth_newBlockFilter
**Request:** `curl -X POST  -H "Content-Type: application/json" --data '{"jsonrpc":"2.0","method":"eth_newBlockFilter","params":[],"id":1}' http://localhost:8545 `
**Response:** `{"jsonrpc":"2.0","id":1,"result":"0x47bf8617648970968c2481aa433c2836"}`

### eth_getFilterChanges
- wait some blocks to be created after create new block filter

**Request:** `curl -X POST  -H "Content-Type: application/json" --data '{"method":"eth_getFilterChanges","params":["0x47bf8617648970968c2481aa433c2836"],"id":1,"jsonrpc":"2.0"}' http://localhost:8545`
**Response:** `{"jsonrpc":"2.0","id":1,"result":["0x44c80c0041d93a5866a448a0d48762f8406e9df6354b099c2469137181d5649a"]}`

### eth_uninstallFilter
```shell
# Uninstall
curl -X POST  -H "Content-Type: application/json" --data '{"method":"eth_uninstallFilter","params":["0x47bf8617648970968c2481aa433c2836"],"id":1,"jsonrpc":"2.0"}' http://localhost:8545 
{"jsonrpc":"2.0","id":1,"result":true}

# Check the result of Uninstall
curl -X POST  -H "Content-Type: application/json" --data '{"method":"eth_uninstallFilter","params":["0x47bf8617648970968c2481aa433c2836"],"id":1,"jsonrpc":"2.0"}' http://localhost:8545 
{"jsonrpc":"2.0","id":1,"error":{"code":-32000,"message":"filter 0x47bf8617648970968c2481aa433c2836 not found"}}
```

### eth_getLogs
**Request:** `curl -X POST  -H "Content-Type: application/json" --data '{"method":"eth_getLogs","params":[{"blockHash":"0x6ebdf8f6a2ccf6c9b9aead31589524949e8e63c16e938f07100725e02c219432"}],"id":1,"jsonrpc":"2.0"}' http://localhost:8545 `
**Result:**
```json
{
  "jsonrpc": "2.0",
  "id": 1,
  "result": [
    {
      "address": "0xb22eAAbf5a3377abDe553f23eA6d82709BA948E0",
      "topics": [
        "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
        "0x000000000000000000000000c6fe5d33615a1c52c08018c47e8bc53646a0e101",
        "0x00000000000000000000000025051c50bb6f302a7bd89100bead641e65da743b"
      ],
      "data": "0x0000000000000000000000000000000000000000000000000000000000000457",
      "blockNumber": "0x2398",
      "transactionHash": "0x80dcea3db9c181748bce112c760102b2f4d6ac7c25d514ec4ed16656b8fefdb6",
      "transactionIndex": "0x1",
      "blockHash": "0x6ebdf8f6a2ccf6c9b9aead31589524949e8e63c16e938f07100725e02c219432",
      "logIndex": "0x0",
      "removed": false
    }
  ]
}
```

### CODE REVIEW

**Comment by zsystm** [comment]
For [CI failure](https://github.com/initia-labs/minievm/actions/runs/9835689661/job/27149847878?pr=28) 

I checked the following local tests passed on my local, but it fails on CI.
- `go test -v ./...`
- `go test ./... -mod=readonly -timeout 12m -race -coverprofile=coverage.txt -covermode=atomic -tags='ledger test_ledger_mock'` (CI ran this command and failed)

Could you please help by re-running the CI tests? 
I suspect there might be an intermittent issue or an environment-specific problem causing the failure.

File: go.mod
Comment (dongsam) [inline go.mod:285]: It seems that the target base branch and main branch of this PR are not yet using the specified version of cometBFT. Since there isn't a separate cometBFT bump PR, it would be helpful to add a more detailed explanation of the reason for the bump in this PR description. [diff c46468756c01...408dc5262680](https://github.com/initia-labs/cometbft/compare/c46468756c01...408dc5262680)

File: jsonrpc/namespaces/eth/filters/filter_system.go
Comment (dongsam) [inline jsonrpc/namespaces/eth/filters/filter_system.go:5]: use code under the GNU license, there may be a requirement that the distribution must also follow the GNU license. However, since minievm currently uses a [non-GNU license](https://github.com/initia-labs/minievm?tab=License-1-ov-file#readme), it is important to clearly state in the PR description that GNU code from evmos/ethermint is being used. This will help the repository maintainer to be aware of this.

**Review by dongsam — COMMENTED** [review]
Aside from the comments regarding the cometBFT bump and the GNU license, LGTM.

File: jsonrpc/namespaces/eth/filters/filter_system.go
Comment (zsystm) [reply jsonrpc/namespaces/eth/filters/filter_system.go:5]: @dongsam 
I have added `License issue` section to the PR description.

> This PR includes some codes that are derived from the Ethermint, which is licensed under the GNU Lesser General Public License (LGPL) v3. Minievm uses the Business Source License (BSL) 1.1, this could potentially lead to a license conflict.

File: go.mod
Comment (zsystm) [reply go.mod:285]: @dongsam 
I have added the reasoning for bumping up cometBFT version.

> updated cometBFT version to [initia-labs/cometbft v0.38.x](https://github.com/initia-labs/cometbft/tree/v0.38.x) which includes [this commit](https://github.com/initia-labs/cometbft/commit/6fe9711db1fc4b8a9d17ae41af183e56480baf77)
without the above commit, broadcasting tx always fails, cannot process any transactions.
check the https://github.com/cometbft/cometbft/issues/3261 for details

File: jsonrpc/namespaces/eth/filters/api.go
Comment (beer-1) [reply jsonrpc/namespaces/eth/filters/api.go:245]: also we don't know eth tx hash at this step, we need to convert cosmos tx to eth tx and get the eth tx hash and use that hash to get receipt.

File: jsonrpc/namespaces/eth/filters/filter_system.go
Comment (beer-1) [reply jsonrpc/namespaces/eth/filters/filter_system.go:5]: This can be problem hmm, I think we can rewrite filter logics to do not use comet but directly subscribe from indexer. 

I will change this part after merging it.

**Review by beer-1 — APPROVED** [review]
LGTM; left some comments but that parts will be changed in next PR, so let's merge it first and I will make new PR to change filter system to do not use comet.

**Pull request merged**
